### PR TITLE
osd/scrub: remove Scrubber sub-objects reliance on PG::gen_prefix()

### PR DESCRIPTION
**Pull request by ronen-fr**
Modify the Scrubber's sub-objects to use their own gen_prefix()
functions, instead of using PG::gen_prefix().

Fixes: https://tracker.ceph.com/issues/51843
Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

**Comment by athanatos** [comment]
Doesn't this skip a ton of information included in PG::gen_prefix()?

**Comment by ronen-fr** [comment]
@athanatos 
> Doesn't this skip a ton of information included in PG::gen_prefix()?

It sure does. But accessing this info when the PG was being removed (deep in the chain of functions that were called
during removal) caused a crash.
And from the short time I spend debugging another problem with these changes in effect - the data is always there, a line or two earlier in the log.

**Comment by athanatos** [comment]
I find that information really valuable.  What access caused the crash?

**Comment by ronen-fr** [comment]
> I find that information really valuable. What access caused the crash?

The specific backtrace in the linked bug:
#7 0x00005638b54f9737 in PG::gen_prefix(std::ostream&) const ()
#8 0x00005638b564e1b4 in std::ostream& _prefix<PG>(std::ostream*, PG*) ()
#9 0x00005638b5652ae9 in PgScrubber::rm_from_osd_scrubbing() ()
#10 0x00005638b566200f in PgScrubber::~PgScrubber() ()
#11 0x00005638b567c3f5 in PrimaryLogScrub::~PrimaryLogScrub() ()
#12 0x00005638b562470d in PrimaryLogPG::~PrimaryLogPG() ()

So - yes, it might be that I could have just stopped using PG:gen_prefix() but still print *pg - I'll
have to check that. But I will be violating the "better not dout in a dtor" "rule".

File: src/osd/pg_scrubber.cc [hunk 1]
```diff
@@ -1889,7 +1901,11 @@ PgScrubber::PgScrubber(PG* pg)
     , m_pg_whoami{pg->pg_whoami}
     , preemption_data{pg}
 {
-  m_fsm = std::make_unique<ScrubMachine>(m_pg, this);
+  std::stringstream prefix;
+  prefix << "osd." << m_osds->whoami << " scrubber pg(" << m_pg_id << "): ";
```
Comment (jdurgin) [inline src/osd/pg_scrubber.cc:1905]: could you use pg[ rather than pg( for all of these prefixes, so that it'll match existing greps for pg-level logging?

**Comment by jdurgin** [comment]
the decoupling approach to avoiding invalid access during destruction seems reasonable to me

**Comment by neha-ojha** [comment]
needs a rebase now

**Review by neha-ojha — APPROVED** [review]
general comment: I think it will be useful to keep the cleanup bits in a separate commit from the bug fix commit, for the sake of backport simplicity

**Comment by ronen-fr** [comment]
Merging based on both my Teuthology runs and on yuriw-2021-11-03_15:48:02-rados-wip-yuri7-testing-2021-11-01-1748-distro-basic-smithi. The Yuri7 failures do not seem to be related:

17 failures:

Might be related to the other two patches:

- 6481465: referred to @ljflores , as there is some chance it relates to #43411;
- 6481640: referred to @myoungwon, as might remotely be related to #43481;
- 6481727: mgr.x core-dumped. I cannot discard relevance to #43411 or #43481. Does not seem related to #42479.  I'd suggest a rerun with current master + patches. Similar test on master (6488752) was OK.
- 6481583 - a timeout. I did not find the cause. Not related to #42479.

The rest:
Also tested and failed in master (http://pulpito.front.sepia.ceph.com/yuriw-2021-11-06_17:01:58-rados-wip-yuri-testing-master-11.5.21-distro-basic-smithi/) (exact same test, or similar enough):
- 6481477 (same message)
- 6481538: similar (not exactly same) to master::6488467
- 6481580: similar (different compression) to master::6488803. Same message.
- 6481608: similar to master::6488543 (the message is different)
- 6481608: similar to master::6488467 (the message is different)
- 6481755: fails the same way on master::6488543
- 6481777: fails on master::6488803 (same message)

A group of '4-fina' tests that either failed or "died":6481690, 6 of the dead
Recurring, old, issues:
 - 6481503 (failure in LibRadosTwoPoolsPP.ManifestSnapRefcount)

Not related:
- 6481528: might be no-deep-scrub related! fails in master, too.
- 6481624: podman core-dumped; does not seem related.
- 6481677: seems to be test-environment issues with access permissions.

**Pull request merged**